src/plugin/task-runner.go

```
@@ -0,0 +1,107 @@
+	"github.com/aws/aws-sdk-go-v2/service/ecs"
+	"github.com/aws/aws-sdk-go-v2/service/ssm"
+)
+
+type TaskRunnerPlugin struct {
+}
+
+type ConfigFetcher interface {
+	Fetch(config *Config) error
+}
+
+func (trp TaskRunnerPlugin) Run(ctx context.Context, fetcher ConfigFetcher) error {
+	var config Config
+	err := fetcher.Fetch(&config)
+	if err != nil {
+		buildkite.LogFailuref("plugin configuration error: %s\n", err.Error())
+		return err
+	}
+
+	buildkite.Log("Executing task-runner plugin\n")
+
+	cfg, err := awsconfig.LoadDefaultConfig(ctx)
+	if err != nil {
+		log.Fatalf("config load failed: %v", err)
+	}
+
+	ssmClient := ssm.NewFromConfig(cfg)
+	log.Printf("Retrieving task configuration from: %s", config.ParameterName)
+	configuration, err := awsinternal.RetrieveConfiguration(ctx, ssmClient, config.ParameterName)
+	if err != nil {
+		log.Fatalf("Failed to retrieve configuration: %v", err)
+	}
+
+	// append Script to configuration.Command. The Script value specifies what script needs to be
+	// executed by the task.
+	configuration.Command = append(configuration.Command, config.Script)
+
+	ecsClient := ecs.NewFromConfig(cfg)
+	taskArn, err := awsinternal.SubmitTask(ctx, ecsClient, configuration)
+	if err != nil {
+		log.Fatalf("Failed to submit task: %v", err)
+	}
+
+	waiterClient := ecs.NewTasksStoppedWaiter(ecsClient, func(o *ecs.TasksStoppedWaiterOptions) {
+		o.MinDelay = time.Second
+		// TODO: This is currently a magic number. If we want this to be configurable, remove the nolint directive and fix it up
+		o.MaxDelay = 10 * time.Second //nolint:mnd
+	})
+	result, err := awsinternal.WaitForCompletion(ctx, waiterClient, taskArn)
+	if err != nil {
+		// TODO: Do we wanna go from fatal, to print, and provide an opportunity to share logs from the task if there any?
+		// That, or we include the log sharing logic within this condition
+		log.Printf("error waiting for task completion: %v", err)
+		log.Fatalf("failure information: %v", result.Failures[0])
```

Return an error rather than logging at fatal. The `main` function should take care of ensuring that the failure is output and that the process returns the correct exit code.

---

Ah, this would be a call out to actually use the `buildkite` loggers in the library. I will go over the items in `Run` now and do a sweep. Thanks!

---

I wanted to bring the specific commit to attention here: [a2732b1](https://github.com/cultureamp/ecs-task-runner-buildkite-plugin/pull/1/commits/a2732b12c608039684b3fa34a3987c8fb6e7d40f). When we get into prototyping we can get a better idea on when to use a group vs a plain ol' log. For now, I opted to use a log.

LMKWYT.